### PR TITLE
Adding support for managed identities to table storage clients

### LINKED ISSUE #9569: Diagnostic Events API should work with a Managed Identity
Currently, the Diagnostic Events API relies on a connection string to read the diagnostic events from table storage. However, if a user has configured their app to use Managed Identity, they will not be able to get these events to help troubleshoot any issues with their function app.

/cc @soninaren @eamonoreilly  @FinVamp1 @fabiocav @mattchenderson

### PR DESCRIPTION
**Pull request by cjaliaga**
### Issue describing the changes in this PR

resolves #9569 

### Pull request checklist

**IMPORTANT**: Currently, changes must be backported to the `in-proc` branch to be included in Core Tools and non-Flex deployments.

* [ ] Backporting to the `in-proc` branch is not required
    * Otherwise: Link to backporting PR
* [ ] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [ ] I have added all required tests (Unit tests, E2E tests)

### CODE REVIEW

File: src/WebJobs.Script.WebHost/Diagnostics/DiagnosticEventTableStorageRepository.cs [hunk 1]
```diff
@@ -41,34 +42,36 @@ public class DiagnosticEventTableStorageRepository : IDiagnosticEventRepository,
         private string _tableName;
 
         internal DiagnosticEventTableStorageRepository(IConfiguration configuration, IHostIdProvider hostIdProvider, IEnvironment environment, IScriptHostManager scriptHostManager,
-            ILogger<DiagnosticEventTableStorageRepository> logger, int logFlushInterval)
+            ILogger<DiagnosticEventTableStorageRepository> logger, AzureComponentFactory componentFactory, AzureEventSourceLogForwarder logForwarder, int logFlushInterval)
```
Comment (cjaliaga) [inline src/WebJobs.Script.WebHost/Diagnostics/DiagnosticEventTableStorageRepository.cs]: I'd prefer here to avoid passing `AzureComponentFactory `and `AzureEventSourceLogForwarder` to create the `TableServiceClientProvider`. Instead we could study (doesn't necessarily need to be in this PR) the creation of a `HostAzureTableStorageProvider` like we do with Blobs: https://github.com/Azure/azure-functions-host/blob/59d8da491072178261050006894f0db0f48951e2/src/WebJobs.Script/StorageProvider/HostAzureBlobStorageProvider.cs 

Is there any implication on not unifying configuration as we do in `HostAzureBlobStorageProvider`?

File: src/WebJobs.Script/WebJobs.Script.csproj [hunk 2]
```diff
@@ -40,6 +40,7 @@
   <ItemGroup>
     <!-- Dependencies needed for Storage Providers -->
     <PackageReference Include="Azure.Core" Version="1.38.0" />
+    <PackageReference Include="Azure.Data.Tables" Version="12.8.3" />
```
Comment (kshyju) [inline src/WebJobs.Script/WebJobs.Script.csproj:43]: The latest version of this package is [12.9.0](https://www.nuget.org/packages/Azure.Data.Tables/12.9.0). Should we use that version?

File: src/WebJobs.Script/StorageProvider/IAzureTableStorageProvider.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace Microsoft.Azure.WebJobs.Script
+{
+    public interface IAzureTableStorageProvider
+    {
+        IConfiguration Configuration { get; }
+
+        bool TryCreateHostingTableServiceClient(out TableServiceClient client);
+
+        bool TryCreateTableServiceClientFromConnection(string connection, out TableServiceClient client);
```
Comment (kshyju) [inline src/WebJobs.Script/StorageProvider/IAzureTableStorageProvider.cs:15]: Should we use the same name for both methods? `TryCreateTableServiceClient`. The second one is just another overload which takes the connection string.

File: src/WebJobs.Script/StorageProvider/IAzureTableStorageProvider.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+using Azure.Data.Tables;
+using Microsoft.Extensions.Configuration;
+
+namespace Microsoft.Azure.WebJobs.Script
+{
+    public interface IAzureTableStorageProvider
+    {
+        IConfiguration Configuration { get; }
+
+        bool TryCreateHostingTableServiceClient(out TableServiceClient client);
```
Comment (kshyju) [inline src/WebJobs.Script/StorageProvider/IAzureTableStorageProvider.cs:13]: Consider adding method summary for these methods/properties.

File: src/WebJobs.Script/StorageProvider/HostAzureTableStorageProvider.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        public HostAzureTableStorageProvider(IConfiguration configuration, ILogger<HostAzureTableStorageProvider> logger, AzureComponentFactory componentFactory, AzureEventSourceLogForwarder logForwarder)
+        {
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _tableServiceClientProvider = new TableServiceClientProvider(componentFactory, logForwarder);
+            Configuration = configuration ?? throw new ArgumentNullException(nameof(configuration));
+        }
+
+        public virtual IConfiguration Configuration { get; private set; }
+
+        public virtual bool TryCreateTableServiceClientFromConnection(string connection, out TableServiceClient client)
```
Comment (kshyju) [inline src/WebJobs.Script/StorageProvider/HostAzureTableStorageProvider.cs:26]: Do we need these methods as virtual? If not, I suggest removing them and marking the class `sealed`.

File: src/WebJobs.Script/StorageProvider/HostAzureTableStorageProvider.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        {
+            var connectionToUse = connection ?? ConnectionStringNames.Storage;
```
Comment (kshyju) [inline src/WebJobs.Script/StorageProvider/HostAzureTableStorageProvider.cs:28]: Since we have the other overload which uses the default ( which is `ConnectionStringNames.Storage`). Should we just throw if `connection` parameter is null here?

File: src/WebJobs.Script/StorageProvider/HostAzureTableStorageProvider.cs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+            try
+            {
+                client = _tableServiceClientProvider.Create(connectionToUse, Configuration);
+                return true;
+            }
+            catch (Exception e)
+            {
+                _logger.LogDebug($"Could not create TableServiceClient. Exception: {e}");
```
Comment (kshyju) [inline src/WebJobs.Script/StorageProvider/HostAzureTableStorageProvider.cs:37]: Consider structured logging.

File: src/WebJobs.Script/StorageProvider/HostAzureTableStorageProvider.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        private readonly TableServiceClientProvider _tableServiceClientProvider;
+
+        public HostAzureTableStorageProvider(IConfiguration configuration, ILogger<HostAzureTableStorageProvider> logger, AzureComponentFactory componentFactory, AzureEventSourceLogForwarder logForwarder)
+        {
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _tableServiceClientProvider = new TableServiceClientProvider(componentFactory, logForwarder);
+            Configuration = configuration ?? throw new ArgumentNullException(nameof(configuration));
+        }
+
+        public virtual IConfiguration Configuration { get; private set; }
```
Comment (kshyju) [inline src/WebJobs.Script/StorageProvider/HostAzureTableStorageProvider.cs:24]: Does this need to be a public property? Looks like it is being used only inside this class. So, we can probably remove it from the interface definition as well?

File: src/WebJobs.Script.WebHost/Diagnostics/DiagnosticEventTableStorageRepository.cs [hunk 9]
```diff
@@ -25,51 +25,48 @@ public class DiagnosticEventTableStorageRepository : IDiagnosticEventRepository,
...
-        public DiagnosticEventTableStorageRepository(IConfiguration configuration, IHostIdProvider hostIdProvider, IEnvironment environment, IScriptHostManager scriptHost,
-            ILogger<DiagnosticEventTableStorageRepository> logger)
-            : this(configuration, hostIdProvider, environment, scriptHost, logger, LogFlushInterval) { }
+        public DiagnosticEventTableStorageRepository(IHostIdProvider hostIdProvider, IEnvironment environment, IScriptHostManager scriptHost, ILogger<DiagnosticEventTableStorageRepository> logger,
+            IAzureTableStorageProvider azureTableStorageProvider)
+            : this(hostIdProvider, environment, scriptHost, logger, azureTableStorageProvider, LogFlushInterval)
+        {
+        }
 
         internal TableServiceClient TableClient
```
Comment (kshyju) [inline src/WebJobs.Script.WebHost/Diagnostics/DiagnosticEventTableStorageRepository.cs:60]: I know this is not the change from your PR, But I am curious why we have this as a property. Looks like it is being used only in this class. I prefer a private variable for this. (With that, we could hopefully remove the logging from the property.)

File: src/WebJobs.Script/StorageProvider/StorageClientProvider.cs [hunk 10]
```diff
@@ -37,7 +37,7 @@ public virtual TClient Create(string name, INameResolver resolver, IConfiguratio
             return this.Create(resolvedName, configuration);
         }
 
-        public virtual TClient Create(string name, IConfiguration configuration)
+        public virtual TClient Create(string name, IConfiguration configuration, TClientOptions options = null)
```
Comment (RohitRanjanMS) [inline src/WebJobs.Script/StorageProvider/StorageClientProvider.cs:40]: `_logForwarder.Start()` might be called twice—once for Blob and once for Table storage. Although it's unlikely because [AddAzureClientsCore](https://github.com/Azure/azure-sdk-for-net/blob/6dbf3c83eae65ae8a11b51fdaf777bf2777cb62b/sdk/extensions/Microsoft.Extensions.Azure/src/AzureClientServiceCollectionExtensions.cs#L54)() registers a singleton instance of AzureEventSourceLogForwarder, did you encounter any duplicate logs from Azure SDKs during your testing? These duplicate logs can be quite bothersome.

File: test/WebJobs.Script.Tests.Integration/Diagnostics/DiagnosticEventTableStorageRepositoryTests.cs [hunk 11]
```diff
@@ -27,9 +30,10 @@ public class DiagnosticEventTableStorageRepositoryTests
         private const string TestHostId = "testhostid";
         private readonly IHostIdProvider _hostIdProvider;
         private readonly TestLoggerProvider _loggerProvider;
+        private readonly IAzureTableStorageProvider _azureTableStorageProvider;
         private IConfiguration _configuration;
         private ILogger<DiagnosticEventTableStorageRepository> _logger;
-        private Mock<IScriptHostManager> _scriptHostMock;
+        private Mock<IScriptHostManager> _scriptHostMock;
```
Comment (RohitRanjanMS) [inline test/WebJobs.Script.Tests.Integration/Diagnostics/DiagnosticEventTableStorageRepositoryTests.cs:36]: Please remove the blank space.

File: src/WebJobs.Script/StorageProvider/HostAzureTableStorageProvider.cs (hunk 7 above)
Comment (RohitRanjanMS) [inline src/WebJobs.Script/StorageProvider/HostAzureTableStorageProvider.cs:37]: Consider logging this as an error.

File: test/WebJobs.Script.Tests.Integration/Storage/TableServiceClientProviderTests.cs [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Xunit;
+
+namespace Microsoft.Azure.WebJobs.Script.Tests.Integration.Storage
+{
+    /// <summary>
+    /// Tests whether the StorageClientProvider can properly create a client and send a request
+    /// </summary>
+    public class TableServiceClientProviderTests
```
Comment (RohitRanjanMS) [inline test/WebJobs.Script.Tests.Integration/Storage/TableServiceClientProviderTests.cs:19]: Consider adding a test case for null/empty connection string.

File: test/WebJobs.Script.Tests.Shared/TestHelpers.cs [hunk 13]
```diff
@@ -513,6 +513,32 @@ public static IServiceCollection AddTestAzureBlobStorageProvider(IServiceCollect
...
+                {
+                    AddTestAzureTableStorageProvider(services, configuration);
+                })
+                .ConfigureAppConfiguration(c =>
+                {
+                    c.AddConfiguration(configuration);
+                })
+                .Build();
+
+            var azureBlobStorageProvider = tempHost.Services.GetRequiredService<IAzureTableStorageProvider>();
```
Comment (RohitRanjanMS) [inline test/WebJobs.Script.Tests.Shared/TestHelpers.cs:529]:
```suggestion
            var azureTableStorageProvider = tempHost.Services.GetRequiredService<IAzureTableStorageProvider>();
```

File: test/WebJobs.Script.Tests.Shared/TestHelpers.cs [hunk 14]
```diff
@@ -513,6 +513,32 @@ public static IServiceCollection AddTestAzureBlobStorageProvider(IServiceCollect
...
+            return azureBlobStorageProvider;
```
Comment (RohitRanjanMS) [inline test/WebJobs.Script.Tests.Shared/TestHelpers.cs:530]:
```suggestion
            return azureTableStorageProvider;
```

File: src/WebJobs.Script.WebHost/Diagnostics/DiagnosticEventTableStorageRepository.cs (hunk 1 above)
Comment (RohitRanjanMS) [reply src/WebJobs.Script.WebHost/Diagnostics/DiagnosticEventTableStorageRepository.cs]: I like this idea. There’s no need to pass AzureEventSourceLogForwarder around since we only need to start it once. We can implement it this way without passing it around.
```
services.AddSingleton<AzureEventSourceLogForwarder>(provider => AzureEventSourceLogForwarder.Start());
```

**Comment by cjaliaga** [comment]
Closed in favor of #10438
